### PR TITLE
MANTA-4860 want smfgen CLI argument support

### PR DESCRIPTION
**Pull request by bahamas10**
a more complete version of https://github.com/joyent/smfgen/pull/6

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -71,36 +71,95 @@ a string in place of an object for the `start`, `stop`, and `refresh` methods.
...
-                </common_name>
-            </template>
-        </service>
-    </service_bundle>
+# CLI Input
+
+The following options can be specified over the CLI:
+
+-  `-d`, `--cwd` becomes `data.start.working_directory`
+-  `-e`, `--env` appends to `data.start.environment`
```
Comment (jlevon) [inline README.md]: nit, should we mention that these also apply to the stop method?
(and refresh?)

File: README.md [hunk 2]
```diff
@@ -71,36 +71,95 @@ a string in place of an object for the `start`, `stop`, and `refresh` methods.
...
+-  `-g`, `--group` becomes `data.start.group`
+-  `-i`, `--ident` becomes `data.ident`
+-  `-l`, `--label` becomes `data.label`
+-  `-p`, `--privilege` appends to `data.start.privileges`
+-  `-S`, `--stop` becomes `data.stop.exec`
+-  `-s`, `--start` becomes `data.start.exec`
+-  `-u`, `--user` becomes `data.start.user`
```
Comment (jlevon) [inline README.md]: I'm curious why timeout, category, dependencies, and refresh aren't supported? In particular "category" seems very useful.

File: README.md (hunk 1 above)
Comment (jlevon) [reply README.md]: Also, given e.g. data.exec and data.timeout works, shouldn't these say that? instead of mentioning start

File: smfgen [hunk 3]
```diff
@@ -21,50 +25,42 @@ function checkAndApplyContext(prefix, conf, applyto)
 	var s = (prefix ? prefix + '.' : '');
 
 	if (conf.hasOwnProperty('user')) {
-		mod_assert.equal(typeof (conf.user), 'string',
-		    '"' + s + 'user" property must be a string');
+		assert.string(conf.user, s + 'user');
```
Comment (jlevon) [inline smfgen]: Don't think I've ever seen old-style assert code before. Wow is assert-plus nicer!

File: smfgen [hunk 4]
```diff
@@ -348,23 +332,112 @@ XmlEmitter.prototype.emitComment = function (content)
...
+		'',
+		'Examples',
+		'  Create a service from a JSON file:',
+		'    $ cat service.json | smfgen',
+		'',
+		'  Create a service from CLI arguments:',
+		'    $ smfgen -i my-service -l "My Service" -s /opt/custom/bin/my-service',
+		'',
+		'Options',
+		'  -d, --cwd <dir>        the directory to use for executing the start method',
```
Comment (jlevon) [inline smfgen]: same comment as before about hard-coding "start" methods

File: smfgen [hunk 5]
```diff
@@ -348,23 +332,112 @@ XmlEmitter.prototype.emitComment = function (content)
...
+		start: {
+			privileges: [],
+			environment: {}
+		},
+		stop: {}
+	};
+	var option;
+	while ((option = parser.getopt()) !== undefined) {
+		switch (option.option) {
+		case 'd': data.start.working_directory = option.optarg; break;
```
Comment (jlevon) [inline smfgen]: see previous comments about hardcoding this as start only

File: smfgen [hunk 6]
```diff
@@ -348,23 +332,112 @@ XmlEmitter.prototype.emitComment = function (content)
...
+		case 'u': data.start.user = option.optarg; break;
+		case 'g': data.start.group = option.optarg; break;
+		case 'h': console.log(usage()); process.exit(0); break;
+		case 'i': data.ident = option.optarg; break;
+		case 'l': data.label = option.optarg; break;
+		case 'e':
+			var arg = option.optarg;
+			var eq = arg.indexOf('=');
+			if (eq < 0)
+				break;
```
Comment (jlevon) [inline smfgen]: shouldn't this error and exit(1) ?

File: README.md (hunk 2 above)
Comment (bahamas10) [reply README.md]: no reason in particular for any options omitted... i can add them in.

File: smfgen (hunk 3 above)
Comment (bahamas10) [reply smfgen]: oh god ha, yeah. it was fun.

File: smfgen (hunk 6 above)
Comment (bahamas10) [reply smfgen]: yes! i'll change that

**Comment by bahamas10** [comment]
It is weird to me seeing things like env vars, privileges, user, group, cwd, etc. applied to the stop and refresh methods... but it doesn't seem wrong or surprisingly honestly.  I think these changes make sense.

File: README.md [hunk 7]
```diff
@@ -112,15 +112,18 @@ This string will be assumed to be the `exec` property of that method.
...
+-  `-e`, `--env` appends to `data.environment`
+-  `-g`, `--group` becomes `data.group`
 -  `-i`, `--ident` becomes `data.ident`
 -  `-l`, `--label` becomes `data.label`
--  `-p`, `--privilege` appends to `data.start.privileges`
+-  `-p`, `--privilege` appends to `data.privileges`
 -  `-S`, `--stop` becomes `data.stop.exec`
 -  `-s`, `--start` becomes `data.start.exec`
--  `-u`, `--user` becomes `data.start.user`
+-  `-r`, `--refresh` becomes `data.refresh.exec`
```
Comment (jlevon) [inline README.md]: nit, this is no longer alphabetically sorted

File: README.md [hunk 8]
```diff
@@ -112,15 +112,18 @@ This string will be assumed to be the `exec` property of that method.
...
+-  `-u`, `--user` becomes `data.user`
```
Comment (jlevon) [inline README.md]: still missing --timeout I think?

File: README.md (hunk 7 above)
Comment (bahamas10) [reply README.md]: I'll ensure that all of it is sorted

File: README.md (hunk 8 above)
Comment (bahamas10) [reply README.md]: ah, good catch. i'll fix that.

**Comment by bahamas10** [comment]
The testing done for this is simply running:

```
$ for f in test/*.js; do echo "$f"; node "$f"; done
test/cli-test.js
test/module-test.js
```

**Pull request merged**